### PR TITLE
[feature](Nereids): push down topN through join

### PR DESCRIPTION
**Pull request by jackwener**
## Proposed changes

Push TopN through Join.

Join Type just can be left/right outer join or cross join, because data of their one child can't be filtered.

new TopN is (original limit + original offset, 0) as limit and offset.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by jackwener** [comment]
run buildall

**Comment by jackwener** [comment]
run buildall

**Comment by jackwener** [comment]
run buildall

**Comment by jackwener** [comment]
run buildall

**Comment by jackwener** [comment]
run buildall

**Comment by jackwener** [comment]
run buildall

**Comment by jackwener** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushdownTopNThroughJoin.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    private Plan pushLimitThroughJoin(LogicalTopN<? extends Plan> topN, LogicalJoin<Plan, Plan> join) {
+        switch (join.getJoinType()) {
+            case LEFT_OUTER_JOIN:
+                Set<Slot> rightOutputSet = join.right().getOutputSet();
+                if (topN.getOrderKeys().stream().map(OrderKey::getExpr)
+                        .anyMatch(e -> Utils.isIntersecting(rightOutputSet, e.getInputSlots()))) {
+                    return null;
+                }
```
Comment (XieJiann) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushdownTopNThroughJoin.java:84]: topN(order by c) -> project(A.a + B.b) -> (A join B)

**Review by keanji-x — APPROVED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushdownTopNThroughJoin.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        );
+    }
+
+    private Plan pushLimitThroughJoin(LogicalTopN<? extends Plan> topN, LogicalJoin<Plan, Plan> join) {
+        switch (join.getJoinType()) {
+            case LEFT_OUTER_JOIN:
+                Set<Slot> rightOutputSet = join.right().getOutputSet();
+                if (topN.getOrderKeys().stream().map(OrderKey::getExpr)
+                        .anyMatch(e -> Utils.isIntersecting(rightOutputSet, e.getInputSlots()))) {
+                    return null;
```
Comment (XieJiann) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PushdownTopNThroughJoin.java:83]: if (join.left().getOutputSet().containsAll(orderbySlots)) {
                    return join.withChildren(topN.withChildren(join.left()), join.right());
}
return null

**Review by morrySnow — APPROVED** [review]

**Pull request merged**